DOC: Improve testing documentation (#31584)
matplotlib/matplotlib: doc/devel/testing.rst

:ref:`QT <backend_dependencies>` is installed.

Writing tests
-------------
Tests are located in :file:`lib/matplotlib/tests`. They are organized to mirror

> the structure of the main code in :file:`lib/matplotlib`. For example, tests for

Reviewer: ```suggestion
the structure of the code in :file:`lib/matplotlib`. For example, tests for
```

the ``mathtext.py`` module are in :file:`lib/matplotlib/tests/test_mathtext.py`.

Naming follows standard pytest conventions:

- files begin with ``"test_"``
- test functions begin with ``"test_"``

> - test class begin ``"Test"``.

Reviewer: ```suggestion
- test classes begin ``"Test"``.
```

We prefer simple test functions, but test classes are also acceptable.
Test function names should be descriptive of what they are testing, and long names
like ``test_to_rgba_array_accepts_color_alpha_tuple_with_multiple_colors()`` are
perfectly fine.

> Simple unit tests

Reviewer: ```suggestion
unit tests
```

the guidance in this section applies whether the unit test is simple or complex

^^^^^^^^^^^^^^^^^

Many elements of Matplotlib can be tested using simple unit tests, e.g. ::

  def test_to_rgba_explicit_alpha_overrides_tuple_alpha():
      assert mcolors.to_rgba(('red', 0.1), alpha=0.9) == (1, 0, 0, 0.9)

> Data in tests

Reviewer: ```suggestion
Data
```

^^^^^^^^^^^^^
Try to use minimal explicit data. Often you can use explicit values like
``[1, 2, 3]``, ``range(5)`` or ``np.arange(5)``. This is cheap and
makes the test more readable.

> 

Reviewer: ```suggestion
Try to use minimal explicit data, such as 
``[1, 2, 3]``, ``range(5)`` or ``np.arange(5)``, because it 
makes the test more readable.

```

When you need more and non-trivial data, generate it programmatically, e.g. ::

   x = np.linspace(0, 2*np.pi, 101)
   y = 2 * np.sin(x) + 1

Use random numbers only, if an algorihmical way to generate the data is not

> possible or too cumbersome. In this case, set the seed to a fixed value to make

Reviewer: ```suggestion
Use random numbers only when an algorithmic way to generate the data is too cumbersome or impossible. In this case, set the seed to a fixed value to make
```

The seed is :ref:`John Hunter's <project_history>` birthday.

Test cleanup
^^^^^^^^^^^^

> We often need figures or modify `.rcParams` to test some functionality. Cleanup

Reviewer: ```suggestion
We often need to create figures or to modify `.rcParams` to test some functionality. Cleanup
```

of such side effects is handled automatically through the the pytest fixture
``matplotlib.testing.conftest.mpl_test_settings``.

In particular, you don't need to call ``plt.close()``.

Testing with figures and Axes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
When you need figures and/or Axes, create them through the standard methods
(``plt.figure()``, ``plt.subplots()``, etc.).

Creating figures and Axes is rather expensive (>100ms). Only create as many as you need for
the test, and reuse them if possible. It is perfectly fine to do multiple operations

> and asserts in one test, e.g.::

Reviewer: Struggling to articulate this, but I think the focus shouldn't be on asserts/operators but on that it's okay to test multiple paramaterizations in one test, which will usually mean multiple asserts. (Even though my bias is definitly for parameterized tests.)

Author: It's not just parametrizations, it could also be related functionality like

```
def test_plot_line_params():
    line, = plt.plot(x, y, color='red')
    assert same_color(line.get_color(), 'red')
  
    line, = plt.plot(x, y, linewidth=10)
    assert line.get_linewith() == 10
```

Proposal:
> It is perfectly fine to test multiple parametrizations or related functionality in one test; i.e. extend the classical test structure *Arrange–Act–Assert* with multiple *Act-Assert* blocks.

  def test_stackplot_facecolor():

>       # Test that facecolors are properly passed and take precedence over colors parameter

Reviewer: ```suggestion
      # Test that facecolors are properly passed and take precedence over the colors parameter
```

Author: I'd rather keep the examples in line with the actual test code. We can make a follow-up to change both. In fact, I only spend limited time to find a suitable example. There may be more concise ones that could be swapped in.

      x = np.linspace(0, 10, 10)
      y1 = 1.0 * x
      y2 = 2.0 * x + 1

      fig, ax = plt.subplots()

      colls = ax.stackplot(x, y1, y2, facecolor=['r', 'b'], colors=['c', 'm'])
      for coll, fcolor in zip(colls, facecolors):

>           assert mcolors.same_color(coll.get_facecolor(), fcolor)

Reviewer: ```suggestion
   assert mcolors.same_color([coll.get_facecolor() for c in colls], facecolors)
   
```
also facecolors isn't defined in this coding block

Author: As opposed to my claim "I'd rather keep the examples in line with the actual test code." I had tried to simplify the code :sheep: Kept the original now. For a follow.up, let's either improve test and doc in a follow up or find a better example.

     # Plural alias should also work
      colls = ax.stackplot(x, y1, y2, facecolors=['r', 'b'], colors=['c', 'm'])
      for coll, fcolor in zip(colls, facecolors):
          assert mcolors.same_color(coll.get_facecolor(), fcolor)

Assert values rather than visual results when feasible. This is clearer,

> less expensive and less fragile than comparing images, e.g. ::

Reviewer: ```suggestion
less computationally expensive and less fragile than comparing images, e.g. ::
```

In general, would be nice to have at the top a sentence or two by what you mean by cheap/expensive b/c I'm fairly certain some of our contributors are iffy on O(n)/O(m) analysis.

  def test_savefig_preserve_layout_engine():
      fig = plt.figure(layout='compressed')
      fig.savefig(io.BytesIO(), bbox_inches='tight')
      assert fig.get_layout_engine()._compress

.. _image-comparison:

> Testing with reference images

Reviewer: ```suggestion
Reference images
``
I'm just not a huge fan of repeating testing in the subtitles when it's in the section name

Author: I think it mostly makes sense here because

- mentioning makes the title more self-contained. If I just see "Data", I have to rember the implicit context. Whereas "Data in tests" says it all.
- there are subtle semantic differences "Testing > Reference images" could also describe where reference images are stored / how they are obtained for running tests. Whereas "Testing with reference images" defines a type of testing, like the previous and next section.

So think the repetition is not purely redundant, but adds value.

Reviewer: I can live w/ that, think it should be "Test data" but that's probably overloaded from the ml context. Not sure if the `ing` verb form or verb form is cleaner here - e.g. "Write tests" vs "Writing  tests"

Author: I'll merge and do a follow up, because this also affects some other places and should be documented in file:///home/tim/git/matplotlib/doc/build/html/devel/style_guide.html
